File: 1.Fundamental-Level/3.OOP/04.Inheritance-and-Abstraction-Homework/1.HumanStudentAndWorker/Student.cs
using System;

namespace _1.HumanStudentAndWorker
{
    public class Student : Human
    {
        private string facultyNumber;

        public Student(string firstName, string lastName, string facultyNumber)
            : base(firstName, lastName)
        {
            this.FacultyNumber = facultyNumber;
        }

        public string FacultyNumber
        {
            get { return this.facultyNumber; }

            set
            {
                if (value.Length < 5 || value.Length > 10)
                {
                    throw new ArgumentException("The faculty number must be between 5 to 10 digits/letters.");
                }

                this.facultyNumber = value;
            }
        }

        public override string ToString()
        {
            return String.Format("{0} {1} with faculty number: {2}",
                    this.FirstName, this.LastName, this.FacultyNumber);
        }
    }
}

File: Java-Development/Java-OOP-Basics/Java-OOP-Basics-Solutions/src/inheritance/_03_Mankind/Student.java
package inheritance._03_Mankind;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Student extends Human {
    private String facultyNumber;

    public Student(String firstName, String lastName, String facultyNumber) {
        super(firstName, lastName);
        this.setFacultyNumber(facultyNumber);
    }

    private void setFacultyNumber(String facultyNumber) {
        Pattern pattern = Pattern.compile("^[a-zA-Z0-9]{5,10}$");
        Matcher matcher = pattern.matcher(facultyNumber);
        if (!matcher.find()) {
            throw new IllegalArgumentException("Invalid faculty number!");
        }

        this.facultyNumber = facultyNumber;
    }

    private String getFacultyNumber() {
        return facultyNumber;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(super.toString());
        sb.append("Faculty number: ").append(this.getFacultyNumber());

        return sb.toString();
    }
}
